feat: improved theme png rendering

NickCraver opened the pull request:
### Prerequisites

- [x] I have read and understood the `CONTRIBUTING` guide
- [x] The commit message follows the [conventional commits][cc] guidelines
- [x] Tests for the changes have been added (for bug fixes/features) (in PR description)
- [x] Docs have been added / updated (for bug fixes/features)

### Description

This is making the following changes to the PNG render pipeline for themes:
1. Use the latest Hack Nerd Font here (from https://github.com/ryanoasis/nerd-fonts/tree/master/patched-fonts/Hack), this solves many to all missing glyph issues.
2. Checks for Nerd Font ranges (defined at https://github.com/ryanoasis/nerd-fonts/wiki/Glyph-Sets-and-Code-Points) and assumes double with for those.

Here are some example before (based on current website) and with this change:

| Theme | Before | After |
| ----- | ------ | ----- |
| M365Princess | ![image](https://user-images.githubusercontent.com/454813/133859221-9bfa3289-662e-459a-90c1-05dff2608f79.png) | ![m365princess](https://user-images.githubusercontent.com/454813/133860124-070dd9aa-21f0-423c-9f77-73846e4626cf.png) |
| agnoster | ![image](https://user-images.githubusercontent.com/454813/133860333-bf41cad6-b714-4c42-b645-be0bdff2799c.png) | ![agnoster](https://user-images.githubusercontent.com/454813/133860322-a83571d7-c2e5-40d0-922b-150e44697e4b.png) |
| blue-owl | ![image](https://user-images.githubusercontent.com/454813/133860580-0b230a5c-7972-4126-9437-143452b861f4.png) | ![blue-owl](https://user-images.githubusercontent.com/454813/133860837-2611cb9e-2d42-4718-9f71-eac3478f71a5.png) |
| bubbles | ![image](https://user-images.githubusercontent.com/454813/133859102-470e6fc5-6e74-4e3e-b6ca-68b0b28ac63b.png) | ![bubbles](https://user-images.githubusercontent.com/454813/133860139-845f9198-d1f5-4a71-8360-3d061ddb0d5d.png) |
| craver | ![image](https://user-images.githubusercontent.com/454813/133858394-4e5fc1d8-1df4-499b-80e9-4ef55e27fbd7.png) | ![craver](https://user-images.githubusercontent.com/454813/133860826-92f8723f-f22d-4a4f-8f8e-9eaee9222ec7.png) |
| darkblood | ![image](https://user-images.githubusercontent.com/454813/133858165-579eca23-e995-47f2-81b1-937ab4c10890.png) | ![darkblood](https://user-images.githubusercontent.com/454813/133858243-14d5a0b3-b814-4178-b4b4-4eefb40b3649.png) |
| material | ![image](https://user-images.githubusercontent.com/454813/133858475-ea8755a2-0dc4-4392-b9c8-19d049e61ef4.png) | ![material](https://user-images.githubusercontent.com/454813/133858441-caad8f15-dd5f-495c-b603-29536e5016bc.png) |
| nu4a | ![image](https://user-images.githubusercontent.com/454813/133858546-a403ae11-a9c4-493d-a26b-3229012d731c.png) | ![nu4a](https://user-images.githubusercontent.com/454813/133860452-b34d516d-11cf-4f78-8ec0-23c8fd4c1bc6.png) |
| paradox | ![image](https://user-images.githubusercontent.com/454813/133858328-e2121433-67ea-492a-a8e6-d576ed41d054.png) | ![paradox](https://user-images.githubusercontent.com/454813/133860460-96c54dc2-4585-4b7e-8ea1-579b804788f2.png) |
| pixelrobots | ![image](https://user-images.githubusercontent.com/454813/133858635-696d700f-65ee-4da8-9a1e-3421e670bc68.png) | ![pixelrobots](https://user-images.githubusercontent.com/454813/133860516-84c45e94-ac7a-43fd-bc57-080bff5fd68c.png) |
| powerlevel10k_classic | ![image](https://user-images.githubusercontent.com/454813/133858718-8410691a-16ab-4669-8d41-631ce11f0f1b.png) | ![powerlevel10k_classic](https://user-images.githubusercontent.com/454813/133860521-5e1a942f-4a93-4ba0-9eb0-de4dc491a206.png) |
| powerlevel10k_rainbow | ![image](https://user-images.githubusercontent.com/454813/133858776-fb57ce3e-42ab-41ca-ab6d-ed0bb2d0751e.png) | ![powerlevel10k_rainbow](https://user-images.githubusercontent.com/454813/133860532-8120a47d-8010-4a34-960a-fb18e41693ff.png) |
| slim | ![image](https://user-images.githubusercontent.com/454813/133858897-2ad6f9c4-9529-4356-863f-a471f548fd35.png) | ![slim](https://user-images.githubusercontent.com/454813/133860195-e2debaf9-1b58-4418-96e8-50557cca1a2e.png) |
| wopian | ![image](https://user-images.githubusercontent.com/454813/133859004-3e9521dd-ca63-4c59-932e-766a9e2f8e12.png) | ![wopian](https://user-images.githubusercontent.com/454813/133859058-30444644-adcc-4121-a455-0ac7d1cbb4eb.png) |

Some things I don't love:
- There may be a wildly simpler way to implement this
- I may have committed atrocities against golang and need to wear the cone of shame
- We're explicitly checking Nerd Font Unicode ranges ([which are documented](https://github.com/ryanoasis/nerd-fonts/wiki/Glyph-Sets-and-Code-Points)), and further excluding a few from that (known to be single width, that I ran into testing)
  - Counter-argument: this is for previews and may rarely or not ever have a negative consequence


...is this a bad idea? I'm not sure - I hope it helps or sparks another idea at least. My feeling is the _result_ is a net win (unless people hate the font!), but the implementation is icky. If the gut reaction to this is "omg Nick, wtf?!?" I don't mind it getting closed one bit, only trying to explore a _possible_ path here to make the themes a bit more accurate in the previews.


[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

NickCraver committed:
feat: improved theme png rendering

Overall this improves glyph rendering in the theme PNGs for docs.
Changes:
- Replaces VictorMono with latest Hack nerd font
- When calculating overall width, count glyphs as double wide
- When rendering a glyph, allow ~1.7x space (this seemed right testing many)

I hope this is a net improvement that helps theme renders be a bit more accurate and shine!

NickCraver committed:
move to 2x for overall width expansion

1.7 introduces some odd offset problems with some backgrounds - 2x is stable.

NickCraver committed:
Based on data from:
https://github.com/ryanoasis/nerd-fonts/wiki/Glyph-Sets-and-Code-Points
...checking those points for nerd font glyphs in play

I'm unfamiliar with Go here and maybe there's a better way to implement,
but the results are much improved.

NickCraver committed:
tweak ranges and bg position for font diff

NickCraver committed:
exclude pixelated (e0c4-e0c7) from doube-width

NickCraver committed:
simplify cyclomatic complexity

Go checks were unhappy - this isn't a hot path so breaking it out
into a format that matches the docs, for easier maintenance.

NickCraver committed:
I have no idea what I'm doing

...but I hope this works - bit simpler layout to maintain too.

JanDeDobbeleer commented on src/image.go:
```diff
@@ -320,9 +372,11 @@ func (ir *ImageRenderer) SavePNG(path string) error {
 		if ir.backgroundColor != nil {
 			dc.SetRGB255(ir.backgroundColor.r, ir.backgroundColor.g, ir.backgroundColor.b)
-			dc.DrawRectangle(x, y-h, w, h+12)
+			dc.DrawRectangle(x, y-h+3, w, h+9)
```
These magic numbers, I'm equally guilty to it, can we give them a name?

JanDeDobbeleer replied:
As it's not "new", we'll keep it as is.

JanDeDobbeleer commented:
@NickCraver this isn't a bad take. To be honest, I was so happy to get this to work in the first place, so this part of the codebase can be pragmatic. Will there be a cleaner way? Probably. Is there a lot of ROI when spending the time on it? Highly unlikely.

JanDeDobbeleer commented on src/image.go:
```diff
@@ -320,9 +372,11 @@ func (ir *ImageRenderer) SavePNG(path string) error {
 		}
 
 		w, h := dc.MeasureString(str)
+		w += (w * float64(ir.runeAdditionalWidth(runes[0])))
```
this is done to double the width or not, right?

NickCraver replied:
Indeed! I tried to go through and add comments on all points. It was light on comments before so wasn't sure on stance there, but hopefully the additions here are good - happy to tweak more!

NickCraver committed:
fix: add comments

Good PR feedback - adding some more commentary for anyone coming across these bits later

NickCraver committed:
typo

NickCraver committed:
Merge branch 'main' into craver/theme-rendering-hack
